faiss_vector_cache.go

```diff
@@ -0,0 +1,195 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build vectors
+// +build vectors
+
+package zap
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	faiss "github.com/blevesearch/go-faiss"
+)
+
+type vecIndexCache struct {
+	closeCh chan struct{}
+	m       sync.RWMutex
+	cache   map[uint16]*cacheEntry
+}
+
+type ewma struct {
+	alpha  float64
+	avg    float64
+	sample uint64
+}
+
+type cacheEntry struct {
+	tracker *ewma
+
+	m     sync.RWMutex
+	index *faiss.IndexImpl
+}
+
+func newVectorIndexCache() *vecIndexCache {
+	return &vecIndexCache{
+		cache:   make(map[uint16]*cacheEntry),
+		closeCh: make(chan struct{}),
+	}
+}
+
+func (vc *vecIndexCache) Clear() {
+	vc.m.Lock()
+	close(vc.closeCh)
+	vc.cache = nil
+	vc.m.Unlock()
+}
+
+func (vc *vecIndexCache) loadVectorIndex(fieldID uint16,
+	indexBytes []byte) (vecIndex *faiss.IndexImpl, err error) {
+	cachedIndex, present := vc.isIndexCached(fieldID)
+	if present {
+		vecIndex = cachedIndex
+		vc.addRef(fieldID)
+	} else {
+		// if the cache doesn't have vector index, just construct it out of the
+		// index bytes and update the cache.
+		vecIndex, err = faiss.ReadIndexFromBuffer(indexBytes, faiss.IOFlagReadOnly)
+		vc.update(fieldID, vecIndex)
+	}
+	return vecIndex, err
+}
+
+func (vc *vecIndexCache) isIndexCached(fieldID uint16) (*faiss.IndexImpl, bool) {
+	vc.m.RLock()
+	entry, present := vc.cache[fieldID]
+	vc.m.RUnlock()
+	if entry == nil {
+		return nil, false
+	}
+
+	entry.m.RLock()
+	rv := entry.index
+	entry.m.RUnlock()
+
+	return rv, present && (rv != nil)
+}
+
+func (vc *vecIndexCache) addRef(fieldIDPlus1 uint16) {
+	vc.m.RLock()
+	entry := vc.cache[fieldIDPlus1]
+	vc.m.RUnlock()
+
+	entry.addRef()
+}
+
+func (vc *vecIndexCache) refresh() (rv int) {
+	vc.m.Lock()
+	cache := vc.cache
+
+	// for every field reconcile the average with the current sample values
+	for fieldIDPlus1, entry := range cache {
+		sample := atomic.LoadUint64(&entry.tracker.sample)
+		entry.tracker.add(sample)
+		// the comparison threshold as of now is (1 - a). mathematically it
+		// means that there is only 1 query per second on average as per history.
+		// and in the current second, there were no queries performed against
+		// this index.
+		if entry.tracker.avg <= (1 - entry.tracker.alpha) {
+			atomic.StoreUint64(&entry.tracker.sample, 0)
+			entry.closeIndex()
+			delete(vc.cache, fieldIDPlus1)
+			continue
+		}
+		atomic.StoreUint64(&entry.tracker.sample, 0)
+	}
+
+	rv = len(vc.cache)
+	vc.m.Unlock()
+	return rv
+}
+
+func (vc *vecIndexCache) monitor() {
+	ticker := time.NewTicker(1 * time.Second)
+	for {
+		select {
+		case <-vc.closeCh:
+			return
+		case <-ticker.C:
+			numEntries := vc.refresh()
+			if numEntries == 0 {
```

Comment:
This can race with update's `if len(vc.cache) == 0 { go vc.monitor() }` below, not sure if this can cause anything deviant though.

Reply:
i dont think it can issues over here, because the updating of the vc.cache (which affects the length check over here and the prev mentioned line) happens under a lock.

if the refresh() function acquired the lock then the update would execute only after the vc.cache has been brought to the latest view (which can affect the length of cache if there were old enough entries being cleaned up in refresh()) 

if update acquired the lock first then the refresh() wouldn't execute until it has the latest view of the vc.cache (in which case there could be an extra entry in the vc.cache). 

however, i think instead of returning the numEntries maybe a bool flag as to whether we exit the monitor or not would make it even more robust, please let me know if that makes sense.